cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStore.java

```diff
@@ -644,7 +645,12 @@ public boolean apply(PluginClass input) {
             if (!row.isEmpty()) {
               addPluginsInRangeToMap(namespace, parentArtifacts, row.getColumns(), plugins, pluginRange, limit);
             }
-            return Collections.unmodifiableSortedMap(plugins);
+            if (limit == Integer.MAX_VALUE || plugins.size() <= limit) {
+              return Collections.unmodifiableSortedMap(plugins);
+            } else {
+              return Collections.unmodifiableSortedMap(
+                plugins.subMap(plugins.firstKey(), ((ArtifactDescriptor) plugins.keySet().toArray()[limit])));
```

Comment:
it's better if we make sure the map does not exceed limit entries instead of filtering at the end. See comment in addPluginsInRangeToMap